src/core/src/op/loop.cpp

```diff
@@ -191,6 +197,86 @@ void op::v5::Loop::validate_and_infer_types() {
     // Body
     m_bodies[0]->validate_nodes_and_infer_types();
 
+    if (!back_edges.empty()) {
+        // if an exact value is available, limit the number of iterations.
+        size_t i, max_num_of_iterations = m_num_iterations == -1 ? INT_MAX : m_num_iterations;
+        bool need_reinvalidate = false;
+        for (i = 0; i < max_num_of_iterations; i++) {
+            need_reinvalidate = false;
+            for (const auto& output_description : m_output_descriptions[0]) {
+                auto body_value = m_bodies[0]->get_results().at(output_description->m_body_value_index)->input_value(0);
+
+                if (auto body_output_description =
+                        ov::as_type_ptr<v0::TensorIterator::BodyOutputDescription>(output_description)) {
+                    if (!back_edges.count(output_description->m_body_value_index))
+                        continue;
+                    const ov::PartialShape& body_value_shape = body_value.get_partial_shape();
+                    auto input_param =
+                        m_bodies[0]->get_parameters().at(back_edges[output_description->m_body_value_index]);
+                    const auto& input_param_ps = input_param->get_partial_shape();
+                    if (body_value_shape.rank().is_static()) {
```

Comment:
I feel like this code could be simplified. Please try the model with while True loop which would Unsqueeze on each iteration -- shape inference would take all MAX_INT iteration to be calculated and it would end up horribly. 
If ranks of previous iteration and back_edged output are different I propose to dynamic rank.

Reply:
I did not get the idea, your means it should change
https://github.com/openvinotoolkit/openvino/blob/932142a995e41dda518fbfa6be0402697b78e626/src/core/src/op/loop.cpp#L202-L204
 to
```c
// do not limit iteration number
// size_t i, max_num_of_iterations = m_num_iterations == -1 ? INT_MAX : m_num_iterations;
bool need_reinvalidate = false;
while (true) {
```
Please help to confirm.

Reply:
The original is dynamic if input and output rank is different, after discussion: https://github.com/openvinotoolkit/openvino/pull/10684#discussion_r921071719
https://github.com/openvinotoolkit/openvino/pull/10684#discussion_r923294675, it seems trying our best to enlarge the estimate shape is another good choice.

Reply:
I think it is a real concern. Let's dig a bit deeper here.

Reply:
Please see the test case. This model could be described like so:
```
input  = Parameter([-1, -1])
while (true) {
   input = unsqueeze(input, 0);
}
```
It must produce dynamic output rank. But with the code from current branch, it doesn't -- it returns `[1,-1,-1]`.
To put it simply, if back edge delivers data of different rank, we must set output shape as dynamic rank.
With all that, code of back edge shape inference could be simplified due to absence of different ranks.

```
TEST(type_prop, loop_operation_dynamic_iter_dynamic_shapes_unsqueeze) {
    // Inner model
    const auto inner_parameter = std::make_shared<opset5::Parameter>(element::dynamic, ov::PartialShape::dynamic());
    const auto unsqueeze = std::make_shared<opset5::Unsqueeze>(
            inner_parameter, opset5::Constant::create(element::i64, {1}, {0}));
    const auto true_const = opset5::Constant::create(element::boolean, {1}, {1});
    auto body = std::make_shared<Function>(OutputVector{unsqueeze, true_const}, ParameterVector{inner_parameter});

    // Outer model
    const auto outer_parameter = std::make_shared<opset5::Parameter>(element::dynamic, ov::PartialShape::dynamic(2));

    const auto trip_count = opset5::Constant::create(element::i64, {1}, {-1});
    const auto execution_condition = opset5::Constant::create(element::boolean, {1}, {1});
    const auto loop = std::make_shared<opset5::Loop>(trip_count, execution_condition);
    loop->set_function(body);
    loop->set_merged_input(inner_parameter, outer_parameter, unsqueeze);
    loop->set_special_body_ports(ngraph::opset5::Loop::SpecialBodyPorts{-1, 1});

    auto outer_result = make_shared<opset5::Result>(loop->get_iter_value(unsqueeze, -1));

    auto outer_model = std::make_shared<Function>(ResultVector{outer_result}, ParameterVector{outer_parameter});
}
```

Also, right now the variable `shape_changed` doesn't react to ranks being different which seems faulty in current solution.

Reply:
Updated including the testcase, thanks.